core/example/benchmark/Cabana_CommPerformance.cpp

```diff
@@ -0,0 +1,573 @@
+            average /= _data[n].size() * comm_size;
+
+            // Output on rank 0.
+            if ( 0 == comm_rank )
+            {
+                file << comm_size << " "
+                     << data_point_vals[n] << " "
+                     << global_min << " "
+                     << global_max << " "
+                     << average << std::endl;
+            }
+        }
+    }
+
+  private:
+
+    MPI_Comm _comm;
+    std::string _name;
+    std::vector<std::size_t> _byte_sizes;
+    std::vector<std::vector<std::chrono::high_resolution_clock::time_point> > _starts;
+    std::vector<std::vector<std::chrono::microseconds> > _data;
+    std::vector<bool> _is_stopped;
+};
+
+//---------------------------------------------------------------------------//
+// Performance test.
+// Data device type is where the data to be communicated lives.
+// Comm device type is the device we want to use for communication.
+template<class DataDevice, class CommDevice>
+void performanceTest( std::fstream& file,
+                      const std::size_t num_particle,
+                      const std::string& test_prefix )
+{
+    // PROBLEM SETUP
+    // -------------
+
+    // Get comm world.
+    auto comm = MPI_COMM_WORLD;
+
+    // Get comm size;
+    int comm_size;
+    MPI_Comm_size( comm, &comm_size );
+
+    // Partition the problem in 3 dimensions.
+    std::vector<int> ranks_per_dim( 3 );
+    MPI_Dims_create( comm_size, 3, ranks_per_dim.data() );
+
+    // Generate a communicator with a cartesian topology and periodic
+    // boundaries.
+    std::vector<int> periodic_dims( 3, 1 );
+    MPI_Comm cart_comm;
+    int reorder_cart_ranks = 1;
+    MPI_Cart_create( comm,
+                     3,
+                     ranks_per_dim.data(),
+                     periodic_dims.data(),
+                     reorder_cart_ranks,
+                     &cart_comm );
+
+    // Get the Cartesian topology index of this rank.
+    int linear_rank;
+    MPI_Comm_rank( cart_comm, &linear_rank );
+    std::vector<int> cart_rank( 3 );
+    MPI_Cart_coords( cart_comm, linear_rank, 3, cart_rank.data() );
+
+    // Compute the 27 ranks this rank communicates with. This includes self
+    // communication. Put self communication first. This list needs to be
+    // unique.
+    std::vector<int> neighbor_ranks;
+    neighbor_ranks.push_back( linear_rank );
+    for ( int k = -1; k < 2; ++k )
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i )
+                if ( !(i==0 && j==0 && k==0) )
+                {
+                    std::vector<int> ncr = { cart_rank[0] + i,
+                                             cart_rank[1] + j,
+                                             cart_rank[2] + k };
+                    int nr;
+                    MPI_Cart_rank( cart_comm, ncr.data(), &nr );
+                    neighbor_ranks.push_back( nr );
+                }
+    std::sort( neighbor_ranks.begin(), neighbor_ranks.end() );
```

Comment:
Is it odd to seemingly do these in a fairly specific order and then call sort? On my first pass I thought when you said `//Put self communication first.` I assumed it was convention, but it's just convenience?

Reply:
oooooh - you found a bug!

Reply:
👐 👐

Reply:
this should be fixed - I made a separate list of unique neighbor ranks that define the pre-computed communication topology.